conformance/tests/httproute-invalid-reference-policy.go

```diff
@@ -0,0 +1,84 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tests
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, HTTPRouteInvalidReferencePolicy)
+}
+
+var HTTPRouteInvalidReferencePolicy = suite.ConformanceTest{
+	ShortName:   "HTTPRouteInvalidReferencePolicy",
+	Description: "A single HTTPRoute in the gateway-conformance-infra namespace should fail to attach to a Gateway in the same namespace if the route has a backendRef Service in the gateway-conformance-app-backend namespace and a ReferencePolicy exists but does not grant permission to route to that specific Service",
+	Features: []suite.SupportedFeature{
+		suite.SupportReferencePolicy,
+	},
+	Manifests: []string{"tests/httproute-invalid-reference-policy.yaml"},
+	Test: func(t *testing.T, s *suite.ConformanceTestSuite) {
+		routeNN := types.NamespacedName{Name: "invalid-reference-policy", Namespace: "gateway-conformance-infra"}
+		gwNN := types.NamespacedName{Name: "same-namespace", Namespace: "gateway-conformance-infra"}
+
+		ns := v1alpha2.Namespace(gwNN.Namespace)
+		kind := v1alpha2.Kind("Gateway")
+
+		t.Run("Route status should have a route parent status with an Accepted condition set to False", func(t *testing.T) {
+			parents := []v1alpha2.RouteParentStatus{{
+				ParentRef: v1alpha2.ParentReference{
+					Group:     (*v1alpha2.Group)(&v1alpha2.GroupVersion.Group),
+					Kind:      &kind,
+					Name:      v1alpha2.ObjectName(gwNN.Name),
+					Namespace: &ns,
+				},
+				ControllerName: v1alpha2.GatewayController(s.ControllerName),
+				Conditions: []metav1.Condition{{
+					Type:   string(v1alpha2.ConditionRouteAccepted),
+					Status: metav1.ConditionFalse,
+				}},
+			}}
+
+			kubernetes.HTTPRouteMustHaveParents(t, s.Client, routeNN, parents, false, 60)
+		})
+
+		t.Run("Gateway should have 0 Routes attached", func(t *testing.T) {
+			gw := &v1alpha2.Gateway{}
+			err := s.Client.Get(context.TODO(), gwNN, gw)
+			require.NoError(t, err, "error fetching Gateway")
+			// There are two valid ways to represent this:
+			// 1. No listeners in status
+			// 2. One listener in status with 0 attached routes
+			if len(gw.Status.Listeners) == 0 {
+				// No listeners in status.
+			} else if len(gw.Status.Listeners) == 1 {
+				require.Equal(t, int32(0), gw.Status.Listeners[0].AttachedRoutes)
+			} else {
+				t.Errorf("Expected no more than 1 listener in status, got %d", len(gw.Status.Listeners))
+			}
```

Comment:
Is there a canonical definition for "Attached"? In this case, if I'm not mistaken, the route should still result in `app-backend-v1` being programmed with 90% weight, with a 10% weight being applied to a 503 response for the `app-backend-v2` backend that's not covered by a ReferencePolicy, based on https://github.com/kubernetes-sigs/gateway-api/blob/master/apis/v1alpha2/httproute_types.go#L191-L196. We (contour) had been considering the route to be "Attached" in this case since it's being at least partially programmed, but I'm not seeing a clear definition one way or another.

Reply:
Agreed that the spec could be more clear on how these situations are intended to be handled - it's potentially more complicated when either resource changes after having been iniitally accepted. I added a discussion topic for the weekly meeting later today to hopefully hear perspectives from various implementations.

Reply:
Opened https://github.com/kubernetes-sigs/gateway-api/issues/1112 for further discussion.

Reply:
@skriss I updated this test to instead check for `ResolvedRefs { status: False, reason: RefNotPermitted }` conditions on both the Route and ListenerStatus, and added a check that HTTP requests eventually start returning 503 status codes - does this match your expected behavior?

Reply:
Definitely makes sense for the HTTPRoute - our implementation is currently using a different reason, but it's wrong & should be switched to `RefNotPermitted`.

We're not currently setting that condition on the Listener status, though - I do see in the API docs that the `RefNotPermitted` listener condition reason says "This reason is used with the “ResolvedRefs” condition when one of the Listener’s Routes has a BackendRef to an object in another namespace, where the object in the other namespace does not have a ReferencePolicy explicitly allowing the reference.", so I guess we should be, though I also wonder why this condition is needed on the Listener (where it could be relevant to any one of the listener's routes) rather than just on the HTTPRoute itself. Looks like this has already been raised in #1112 (specifically in https://github.com/kubernetes-sigs/gateway-api/issues/1112#issuecomment-1099655762).

Reply:
Thanks for the detail @skriss! I skipped the listener check for now to hopefully get this merge-able and can follow up on that later once we can reach consensus across implementations on expected behavior.